src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

```diff
@@ -0,0 +1,491 @@
+     */
+    public boolean isUpdateGridEnabled()
+    {
+        return elementCache().updateGridButton.isEnabled();
+    }
+
+    @Override
+    protected ElementCache elementCache()
+    {
+        return (ElementCache) super.elementCache();
+    }
+
+    @Override
+    protected ElementCache newElementCache()
+    {
+        return new ElementCache();
+    }
+
+    protected class ElementCache extends ModalDialog.ElementCache
+    {
+        protected final Locator contentPanelLocator = Locator.byClass("field-modal__col-content");
+
+        // This is present to items in both panels.
+        protected final Locator listItemName = Locator.tagWithClass("div", "field-name");
+
+        protected final WebElement undoEditsButton = Locator.tagWithText("span", "Undo edits")
+                .refindWhenNeeded(this);
+
+        // The checkbox has no id or name.
+        protected final Checkbox checkbox = new Checkbox(Locator.tagWithAttribute("input", "type", "checkbox")
+                .findWhenNeeded(this));
+
+        protected final WebElement updateGridButton = Locator.button("Update Grid")
+                .findWhenNeeded(this);
+
+        // The 'pencil' to edit a label. Only in the Shown in Grid panel.
+        protected final WebElement fieldLabelEdit = Locator.tagWithClass("input", "form-control")
+                .refindWhenNeeded(shownInGridPanel());
+
+        protected WebElement availableFieldsPanel()
+        {
+            return contentPanelLocator.findElements(this).get(0);
+        }
+
+        protected WebElement shownInGridPanel()
+        {
+            return contentPanelLocator.findElements(this).get(1);
+        }
+
+        // Will get all the list items that match the fieldName.
+        protected List<WebElement> getListItemElements(WebElement panel, String fieldName)
+        {
+            return Locator.tagWithClass("div", "list-group-item")
+                    .withDescendant(Locator.tagWithClass("div", "field-name").withText(fieldName))
+                    .findElements(panel);
+        }
+
+        // Will get the first list item that matches the fieldName.
+        protected WebElement getListItemElement(WebElement panel, String fieldName)
+        {
+            return Locator.tagWithClass("div", "list-group-item")
+                    .withDescendant(Locator.tagWithClass("div", "field-name").withText(fieldName))
+                    .findElement(panel);
+        }
+
+        // The data-fieldkey attribute is only present in items in the Available Fields panel.
+        // Similar value to field-name (no spaces, but casing is the same). For child fields it will contain the parent path.
+        protected WebElement getListItemElementByFieldKey(WebElement panel, String fieldKey)
+        {
+            return Locator.tagWithClass("div", "list-group-item")
+                    .withAttribute("data-fieldkey", fieldKey)
+                    .findElement(panel);
+        }
```

Comment:
Why specify the panel if this is only relevant to the available fields panel?
```suggestion
        protected WebElement getListItemElementByFieldKey(WebElement panel, String fieldKey)
        {
            return Locator.tagWithClass("div", "list-group-item")
                    .withAttribute("data-fieldkey", fieldKey)
                    .findElement(availableFieldsPanel());
        }
```

Reply:
Agreed. Changed.